roles/branch-validation/tasks/ha.yml

```diff
@@ -1,114 +1,105 @@
+    - "{{ (groups[instance_group][1:]|difference(groups[ha_instance_group]))[0] }}"
+  when: not converged
 
 - name: floating ips are reachable
   command: "ping -c 5 {{ item.floating_ip.floating_ip_address }}"
-  with_items: fip_controllers.results + fip_compute.results
-  changed_when: false
+  with_items: fip_compute.results
   register: pings
   until: pings|success
-  delay: 1
+  delay: 10
   retries: 5
+  sudo: no
   delegate_to: localhost
+  when: not converged
 
-- name: define ha master
-  set_fact: master=true
+- name: define master
+  set_fact:
+    master: true
   when: floating_ip in ansible_all_ipv4_addresses
 
+- name: define secondary
+  set_fact:
+    master: false
+  when: floating_ip not in ansible_all_ipv4_addresses
+
 # initiate hard reboot on the current master controller
-- name: initiate a failover on controller 0
-  shell: echo 'b' | tee /proc/sysrq-trigger
-  when: master is defined
+- name: reboot controller 0 (master)
+  shell: shutdown -r now
+  when: master
 
-- name: initiate a failover on controller 1
+- name: wait for the panic controller 0 (master)
+  wait_for:
+    host: "{{ inventory_hostname }}"
+    port: 22
+    state: stopped
+    delay: 10
+    timeout: 30
   delegate_to: "{{ groups[ha_instance_group][1] }}"
-  shell: echo 'b' | tee /proc/sysrq-trigger
-  when: master is not defined
-
-- name: wait for the panic
-  wait_for: "host={{ inventory_hostname }} port=22 state=stopped delay=10 timeout=30"
-  when: master is defined
-  delegate_to: localhost
-
-- name: wait for the panic
-  wait_for: "host={{ groups[ha_instance_group][1] }} port=22 state=stopped delay=10 timeout=30"
-  when: master is not defined
-  delegate_to: localhost
+  when: master
 
-- name: wait for the node to boot
-  pause: seconds=30
-
-- name: check non-controller instance connectivity by way of floating IPs
-  command: "ping -c 5 {{ fip_compute.results[0].floating_ip.floating_ip_address }}"
-  changed_when: false
-  register: pings
-  until: pings|success
-  delay: 1
-  retries: 5
-  delegate_to: localhost
-
-- name: check remaining controller instance connectivity
-  command: "ping -c 5 {{ fip_controllers.results[0].floating_ip.floating_ip_address }}"
-  changed_when: false
-  register: pings
-  until: pings|success
-  delay: 1
-  retries: 5
-  when: master is not defined
-  delegate_to: localhost
+- name: reboot controller 1 (secondary)
+  shell: shutdown -r now
+  delegate_to: "{{ groups[ha_instance_group][1] }}"
+  when: not master
 
-- name: check remaining controller instance connectivity
-  command: "ping -c 5 {{ fip_controllers.results[1].floating_ip.floating_ip_address }}"
-  changed_when: false
-  register: pings
-  until: pings|success
-  delay: 1
-  retries: 5
-  delegate_to: localhost
-  when: master is defined
+- name: wait for the panic controller 1 (secondary)
+  wait_for:
+    host: "{{ groups[ha_instance_group][1] }}"
+    port: 22
+    state: stopped
+    delay: 10
+    timeout: 30
+  when: not master
```

Comment:
If I read this right, you're rebooting one node and as soon as it's "down" you're rebooting the other node. That's not what we want to do.

We want to stop one node, the primary node. Once it's stopped, we want to validate that the API is still up and that instances are still reach able (skipping any instances that may have been on that node had it been a converged cluster).

Then we want to bring that node back up, and when it's up, check that API is still up and that instances are all reachable.

Bonus points for rebooting the secondary node and doing the same tests.